### PR TITLE
[HiGHS] Added new port for the HiGHS optimization library

### LINKED ISSUE #30527: [New Port Request] highs
**Library name:** highs

**Library description:**  HiGHS is a high-performance optimization solver for linear, mixed-integer, and quadratic optimization problems.

**Source repository URL:**  https://github.com/ERGO-Code/HiGHS

**Project homepage:** https://highs.dev/

**Notes:**

- I've already written a port that I've tested on windows and linux, but I wanted to check to see if it's suitable to submit a PR.  
- HiGHS can be used as an application or a library, the port I'm suggesting is only for the library.
- I'm not a developer on HiGHS, but am happy to reach out to the owners if needed.

Thanks!

### PR DESCRIPTION
**Pull request by mathgeekcoder**
Fixes #30527

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/highs/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,23 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO ERGO-Code/HiGHS
+    REF v1.5.1
```
Comment (JonLiu1993) [inline ports/highs/portfile.cmake:4]:
```suggestion
    REF "v${VERSION}"
```

File: ports/highs/vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "highs",
+  "version": "1.5.1",
+  "description": "High performance library to solve linear, mixed-integer, and convex quadratic optimization problems.",
+  "homepage": "https://highs.dev",
```
Comment (JonLiu1993) [inline ports/highs/vcpkg.json:5]:
```suggestion
  "homepage": "https://highs.dev",
  "license": "MIT",
```

**Review by JonLiu1993 — DISMISSED** [review]

**Comment by JonLiu1993** [comment]
Tested usage by `highs:x64-windows` successfully:
````
The package highs is compatible with built-in CMake targets:

    find_package(highs REQUIRED)
    target_link_libraries(main PRIVATE highs::highs)
````

File: ports/highs/fix-threads.patch [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+-target_link_libraries(highs Threads::Threads)
+-else()
+-target_link_libraries(libhighs Threads::Threads)
+-endif()
++#find_package(Threads REQUIRED)
++#if (FAST_BUILD)
++#target_link_libraries(highs Threads::Threads)
++#else()
++#target_link_libraries(libhighs Threads::Threads)
++#endif()
```
Comment (dg0yt) [inline ports/highs/fix-threads.patch:20]: Why is this completely disabled now?

File: ports/highs/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS -DFAST_BUILD=ON
+)
+
+vcpkg_cmake_install()
+vcpkg_cmake_config_fixup(CONFIG_PATH "lib/cmake/highs")
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include" "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig" "${CURRENT_PACKAGES_DIR}/lib/pkgconfig")
```
Comment (dg0yt) [inline ports/highs/portfile.cmake:21]: Can't pkgcofig be fixed?

File: ports/highs/vcpkg.json [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  "license": "MIT",
+  "supports": "!arm64",
```
Comment (dg0yt) [inline ports/highs/vcpkg.json:7]: Upstream has binaries for arm64-osx.

File: ports/highs/fix-threads.patch (hunk 3 above)
Comment (mathgeekcoder) [reply ports/highs/fix-threads.patch]: It was causing issues on windows when actually trying to use the library.  I've updated the code to still use threads, but instead make it more compatible.

File: ports/highs/portfile.cmake (hunk 4 above)
Comment (mathgeekcoder) [reply ports/highs/portfile.cmake]: Fair point.  My original "fix" was lazy.  I've improved it now.

File: ports/highs/vcpkg.json (hunk 5 above)
Comment (mathgeekcoder) [reply ports/highs/vcpkg.json]: Again, good point.  Fixed.  Took a little digging to find compile issue with arm64-windows.

File: ports/highs/fix-threads.patch [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+@@ -792,7 +792,7 @@ endif()
+ 
+ find_package(Threads REQUIRED)
+ if (FAST_BUILD)
+-target_link_libraries(highs Threads::Threads)
++target_link_libraries(highs ${CMAKE_THREAD_LIBS_INIT})
+ else()
+-target_link_libraries(libhighs Threads::Threads)
++target_link_libraries(libhighs ${CMAKE_THREAD_LIBS_INIT})
+ endif()
```
Comment (dg0yt) [inline ports/highs/fix-threads.patch:14]: I wonder why this change should be needed.
If it is missing from exported config, `find_dependency(Threads)` (plus including of `CMakeFindDependencyMacro`) must be added there. (This should also be upstreamed.)

File: ports/highs/fix-pkgconfig-path.patch [hunk 7]
```diff
@@ -0,0 +1,10 @@
+diff --git a/highs.pc.in b/highs.pc.in
+index d962ddffa..da7df76f0 100644
+--- a/highs.pc.in
++++ b/highs.pc.in
+@@ -1,4 +1,4 @@
+-prefix=@CMAKE_INSTALL_PREFIX@
++prefix=${pcfiledir}/../..
+ libdir=${prefix}/@CMAKE_INSTALL_LIBDIR@
+ includedir=${prefix}/@CMAKE_INSTALL_INCLUDEDIR@/highs
+
```
Comment (dg0yt) [inline ports/highs/fix-pkgconfig-path.patch:10]: Instead of this patch, a call to `vcpkg_fixup_pkgconfig` should be added to the portfile.

File: ports/highs/portfile.cmake [hunk 8]
```diff
@@ -0,0 +1,24 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO ERGO-Code/HiGHS
+    REF "v${VERSION}"
+    SHA512 ce2003f7328d49e2292135b740204eb485c8c6694fa017c6604e6d4057913d5d5eea517561ffe4b7251b0fbe230bc8288ab6d91dd9432ffc96e792a1ce866424
+    HEAD_REF master
+      PATCHES
```
Comment (dg0yt) [inline ports/highs/portfile.cmake:7]: Please align `PATCHES` with `HEAD_REF`, and indent the actual patches four spaces deeper.

File: ports/highs/portfile.cmake [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+      fix-pkgconfig-path.patch
+      fix-hconfig-path.patch
+      fix-cmake-output.patch
+      fix-threads.patch
+)
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS -DFAST_BUILD=ON
```
Comment (dg0yt) [inline ports/highs/portfile.cmake:16]: Please format as
~~~
    OPTIONS
        -DFAST_BUILD=ON
~~~

**Review by dg0yt — COMMENTED** [review]
The port looks good now, but I do have some more comments.

File: ports/highs/fix-pkgconfig-path.patch (hunk 7 above)
Comment (mathgeekcoder) [reply ports/highs/fix-pkgconfig-path.patch]: Great spotting. Thanks! I wasn't aware of this command.

File: ports/highs/portfile.cmake (hunk 8 above)
Comment (mathgeekcoder) [reply ports/highs/portfile.cmake]: Whoops.  Thanks for catching this!

File: ports/highs/portfile.cmake (hunk 9 above)
Comment (mathgeekcoder) [reply ports/highs/portfile.cmake]: Done!

File: ports/highs/fix-threads.patch (hunk 6 above)
Comment (mathgeekcoder) [reply ports/highs/fix-threads.patch:14]: Sorry, I'm not exactly sure what you're suggesting here.  Do the consumers of the library need to include a ```find_dependency(Threads)``` in their CMakeLists.txt?  FYI: This solution technically works, but I didn't think it was "best practice".

I'd love to have a cleaner fix for this, so if you have ideas - please let me know!

Comment (mathgeekcoder) [reply ports/highs/fix-threads.patch:14]: Actually, I think I might know what you mean.  I'll try something, please let me know if it's better!
